### PR TITLE
Updates, clippy, fmt

### PR DESCRIPTION
**Pull request by Dzordzu**

### CODE REVIEW

**Review by balamurugana — COMMENTED** [review]
Thanks for the PR. LGTM. Please add `clippy` into Github workflow https://github.com/minio/minio-rs/blob/master/.github/workflows/rust.yml

**Comment by Dzordzu** [comment]
Still loads of clippy warnings were present. Most of them were about large variants. That would require the library to wrap most things in boxes (obviously it's a breaking change). Furthermore some methods were noted with `#[allow(clippy::too_many_arguments)]` (again redesign…). I've also fixed some more issues. Now it's ready for the clippy integration :)

**Comment by Dzordzu** [comment]
Last commit won't work. I need a little time

**Comment by Dzordzu** [comment]
Ok. Should be ready (I guess)

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -8,25 +8,31 @@ on:
...
 
     runs-on: ubuntu-latest
-
     steps:
-    - uses: actions/checkout@v3
-    - name: Build and Check style
-      run: |
-        cargo build --verbose
-        cargo fmt --all -- --check
+      - uses: actions/checkout@v3
```
Comment (balamurugana) [inline .github/workflows/rust.yml:18]: Avoid unnecessary diffs and make sure you do minimal changes.

**Comment by Dzordzu** [comment]
Okey. Due to the issues shown in the [this commit](https://github.com/dna2fork/tensorboard/commit/5d560de192ce2b73a05535fbc8ee8f2f0c4a2221), I got back to the simple command run. Sorry for this mess, I never used gh runner for the rust projects

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: .github/workflows/rust.yml (hunk 1 above)
Comment (Dzordzu) [reply .github/workflows/rust.yml:18]: > Avoid unnecessary diffs 

Hm?

> make sure you do minimal changes.

It seemed to be a minor change. Well 
image

Comment (Dzordzu) [reply .github/workflows/rust.yml:18]: Ehs. I may be a little tired…

**Comment by Dzordzu** [comment]
That should do it. But check it locally. I've been working for 10h today. My weekend was similar. Sorry for the mess with pipelines. Ensure that changes are fine and start the workflow.

image

File: .github/workflows/rust.yml (hunk 1 above)
Comment (balamurugana) [reply .github/workflows/rust.yml:19]: We do not need `yamllint` . Please add what is really needed here.

File: .test.env [hunk 2]
```diff
@@ -0,0 +1,7 @@
+export SERVER_ENDPOINT=localhost:9000
+export ACCESS_KEY="masoud" 
+export SECRET_KEY="Strong#Pass#2022" 
+export ENABLE_HTTPS=1 
+export IGNORE_CERT_CHECK=0 
+export SSL_CERT_FILE=./tests/public.crt 
+export MINIO_CI_CD=true
```
Comment (balamurugana) [inline .test.env:7]: We do not need anything like this.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,7 @@
+test-docker-compose-nextest:
+	cd ./tests && docker-compose down && docker-compose up -d
+	source ./.test.env && cargo nextest run --release
+
+test-docker-compose:
+	cd ./tests && docker-compose down && docker-compose up -d
+	source ./.test.env && cargo test --release
```
Comment (balamurugana) [inline Makefile:7]: We do not need anything like this.

File: tests/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+      MINIO_ROOT_PASSWORD: Strong#Pass#2022
+      MINIO_ROOT_USER: masoud
+      MINIO_CI_CD: true
+      MINIO_NOTIFY_WEBHOOK_ENABLE_miniorustest: on
+      MINIO_NOTIFY_WEBHOOK_ENDPOINT_miniorustest: http://example.org/
+
+    volumes:
+      - ./private.key:/certs/private.key
+      - ./public.crt:/certs/public.crt
+    command: server /data/test-xl/{1...4} --certs-dir /certs
```
Comment (balamurugana) [inline tests/docker-compose.yml:20]: We do not need anything like this.

File: tests/start-server.sh [hunk 5]
```diff
@@ -1,16 +1,29 @@
...
-cp ./tests/public.crt ./tests/private.key /tmp/certs/
+cp $CURR_DIR/tests/public.crt $CURR_DIR/tests/private.key /tmp/certs/
 
 (MINIO_CI_CD=true \
-    MINIO_NOTIFY_WEBHOOK_ENABLE_miniojavatest=on \
-    MINIO_NOTIFY_WEBHOOK_ENDPOINT_miniojavatest=http://example.org/ \
-    ./minio server /tmp/test-xl/{1...4}/ --certs-dir /tmp/certs/ &)
+    MINIO_NOTIFY_WEBHOOK_ENABLE_miniorustest=on \
+    MINIO_NOTIFY_WEBHOOK_ENDPOINT_miniorustest=http://example.org/ \
+    $CURR_DIR/minio server /tmp/test-xl/{1...4}/ --certs-dir /tmp/certs/ &)
```
Comment (balamurugana) [inline tests/start-server.sh:27]: We do not need anything like this.

File: tests/tests.rs [hunk 6]
```diff
@@ -1229,6 +1189,7 @@ async fn s3_tests() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
     println!("compose_object()");
     ctest.compose_object().await;
 
+    // bucket notification fail
```
Comment (balamurugana) [inline tests/tests.rs:1192]: Remove this.

File: tests/tests.rs [hunk 7]
```diff
@@ -1171,9 +1127,13 @@ async fn s3_tests() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
...
     let secret_key = std::env::var("SECRET_KEY")?;
-    let secure = std::env::var("ENABLE_HTTPS").is_ok();
+    let secure = !["no", "false", "0"]
+        .into_iter()
+        .any(|x| std::env::var("ENABLE_HTTPS").unwrap_or("no".into()) == x);
     let ssl_cert_file = std::env::var("SSL_CERT_FILE")?;
-    let ignore_cert_check = std::env::var("IGNORE_CERT_CHECK").is_ok();
+    let ignore_cert_check = !["no", "false", "0"]
+        .into_iter()
+        .any(|x| std::env::var("IGNORE_CERT_CHECK").unwrap_or("no".into()) == x);
```
Comment (balamurugana) [inline tests/tests.rs:1136]: There are many options if you deal with on/off flag. Just leave it to current way.

**Review by balamurugana — CHANGES_REQUESTED** [review]

**Comment by balamurugana** [comment]
> Do you have any other requests for changes?

Just reviewed and left comments.

**Comment by Dzordzu** [comment]
For the rest of the comments:

As I noted in #31  `/tmp/` directory may cause some issues with the MinIO. Particularly, I got 

```
ERROR Unable to initialize backend: Unable to write to the backend
      > Please ensure your drive supports O_DIRECT
      HINT:
        Drives do not support O_DIRECT flags, MinIO erasure coding requires filesystems with O_DIRECT support
```

It's noted as [intended bahavior](https://github.com/minio/minio/issues/10277). All of these changes are meant to simplify process of the local development (especially testing). This will certainly shorten some actions

### TLDR;
The command provided in the workflows does not work in many local environments. It would be appropriate to give tools to run local tests without need to copy-paste commands

File: tests/start-server.sh (hunk 5 above)
Comment (Dzordzu) [reply tests/start-server.sh:27]: https://github.com/minio/minio-rs/pull/32#issuecomment-1578311770

File: tests/docker-compose.yml (hunk 4 above)
Comment (Dzordzu) [reply tests/docker-compose.yml:20]: https://github.com/minio/minio-rs/pull/32#issuecomment-1578311770

File: .test.env (hunk 2 above)
Comment (Dzordzu) [reply .test.env:7]: https://github.com/minio/minio-rs/pull/32#issuecomment-1578311770

File: tests/start-server.sh (hunk 5 above)
Comment (Dzordzu) [reply tests/start-server.sh:27]: Actually, I can remove that

File: .github/workflows/rust.yml [hunk 8]
```diff
@@ -1,3 +1,4 @@
+---
```
Comment (balamurugana) [inline .github/workflows/rust.yml:1]: Please remove this.

File: tests/tests.rs (hunk 7 above)
Comment (balamurugana) [reply tests/tests.rs:1136]: These env variables also compatible with our `mint` testing project. I would prefer to leave them as is.

File: .test.env (hunk 2 above)
Comment (balamurugana) [reply .test.env:7]: How you run your S3 server/service for local testing is out of scope of this project. You cannot use Github CI script locally. In fact, there is no need to test `minio-rs` locally. If a contributor does that, it is better to add respective functional tests rather than testing locally.

We need to keep the project tidy.

File: tests/start-server.sh [hunk 9]
```diff
@@ -9,8 +9,8 @@ mkdir -p /tmp/certs
 cp ./tests/public.crt ./tests/private.key /tmp/certs/
 
 (MINIO_CI_CD=true \
-    MINIO_NOTIFY_WEBHOOK_ENABLE_miniojavatest=on \
-    MINIO_NOTIFY_WEBHOOK_ENDPOINT_miniojavatest=http://example.org/ \
+    MINIO_NOTIFY_WEBHOOK_ENABLE_miniorustest=on \
+    MINIO_NOTIFY_WEBHOOK_ENDPOINT_miniorustest=http://example.org/ \
```
Comment (balamurugana) [inline tests/start-server.sh:13]: This will not work in our `mint` testing i.e. the test server does not have this webhook and notification target. You could leave as is here and below `tests/tests.rs`

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: tests/start-server.sh (hunk 9 above)
Comment (Dzordzu) [reply tests/start-server.sh:13]: > How you run your S3 server/service for local testing is out of scope of this project

Nope. It's not out of scope. It's extremely in the scope

> You cannot use Github CI script locally

But I can copy commands from there and emulate github actions

>  In fact, there is no need to test minio-rs locally

New features? Fixing/changing/replacing internal methods? As a contributor I would like to have an easy tool to test new changes. Furthermore, it's just 3 files. It reduces contributors time spent on introduction to repository and code testing by the large factor. It's also saving time of the minio team - you don't have to spend time allowing workflows that will obviously fail

> If a contributor does that, it is better to add respective functional tests rather than testing locally.

How does one connect to another in this case?

**Comment by balamurugana** [comment]
Already addressed by https://github.com/minio/minio-rs/pull/38
